Update optional client cert doc for 6.0 features (#23226)
dotnet/AspNetCore.Docs: aspnetcore/security/authentication/certauth.md

::: moniker-end

ASP.NET Core 5 and later adds more convenient support for redirecting to aquire optional client certificates. For more information, see the [Optional certificates sample](https://github.com/dotnet/aspnetcore/tree/9ce4a970a21bace3fb262da9591ed52359309592/src/Security/Authentication/Certificate/samples/Certificate.Optional.Sample).

* For requests to the web app that require a client certificate and don't have one:
  * Redirect to the same page using the client certificate protected subdomain.
  * For example, redirect to `myClient.contoso.com/requestedPage`. Because the request to `myClient.contoso.com/requestedPage` is a different hostname than `contoso.com/requestedPage`, the client establishes a different connection and the client certificate is provided.
  * For more information, see <xref:security/authorization/introduction>.

&dagger; Server Name Indication (SNI) is a TLS extension to include a virtual domain as a part of SSL negotiation. This effectively means the virtual domain name, or a hostname, can be used to identify the network end point.

### Renegotiation

> TLS renegotiation is a process by which the client and server can re-assess the encryption requirments for an individual connection, including requesting a client certificate if not previously provided. This is not recommended because:

Reviewer: ```suggestion
TLS renegotiation is a process by which the client and server can re-assess the encryption requirements for an individual connection, including requesting a client certificate if not previously provided. TLS renegotiation is a security risk and isn't recommended because:
```

- In HTTP/1.1 the server must first buffer or consume any HTTP data that is in flight such as POST request bodies to make sure the connection is clear for the renegotiation. Otherwise the renegotiation can hang or fail.
- HTTP/2 and HTTP/3 [explicitly prohibit](https://tools.ietf.org/html/rfc7540#section-9.2.1) renegotiation.
- There are security risks associated with renegotiation. TLS 1.3 removed renegotiation of the whole connection and replaced it with a new extension for requesting only the client certificiate after the start of the connection. This mechanism is exposed via the same APIs and is still subject to the prior constraints of buffering and HTTP protocol versions.

The implmentation and configuration of this feature varies by server and framework version.

#### IIS

> IIS manages the client certificate negotation on your behalf. A subsection of the application can enable the `SslRequireCert` option to negotiate the client certificate for those requests. See the [IIS docs](/iis/configuration/system.webserver/security/access#configuration) for details.

Reviewer: ```suggestion
IIS manages the client certificate negotiation on your behalf. A subsection of the application can enable the `SslRequireCert` option to negotiate the client certificate for those requests. See [Configuration in the IIS documentation](/iis/configuration/system.webserver/security/access#configuration) for details.
```

> IIS will automatically buffer any request body data up to a configured size limit before renegotiating. Requests that exceed the limit will be rejected with a 413 response. This limit defaults to 48MB and is configurable by setting the [uploadReadAheadSize](/iis/configuration/system.webserver/serverruntime).

Reviewer: ```suggestion
IIS will automatically buffer any request body data up to a configured size limit before renegotiating. Requests that exceed the limit are rejected with a 413 response. This limit defaults to 48MB and is configurable by setting the [uploadReadAheadSize](/iis/configuration/system.webserver/serverruntime).
```

*NOTE* The behavior of the `ClientCertificate` property changed in .NET 6, see https://github.com/aspnet/Announcements/issues/466 if working with prior versions.

::: moniker-end

::: moniker range="< aspnetcore-6.0"

There is a [known issue](https://github.com/dotnet/aspnetcore/issues/33586) where enabling `AllowRenegotation` can cause the renegotiation to happen synchronously when accessing the <xref:Microsoft.AspNetCore.Http.ConnectionInfo.ClientCertificate> property. Call the <xref:Microsoft.AspNetCore.Http.ConnectionInfo.GetClientCertificateAsync*> method to avoid this. This has been addressed in .NET 6, see https://github.com/aspnet/Announcements/issues/466. Note `GetClientCertificateAsync` can return a null certificiate if the client declines to provide one.

::: moniker-end

#### Kestrel

Kestrel controls client certificate negotation with the <xref:Microsoft.AspNetCore.Server.Kestrel.Https.HttpsConnectionAdapterOptions.ClientCertificateMode> option.

::: moniker range=">= aspnetcore-6.0"

> <xref:Microsoft.AspNetCore.Server.Kestrel.Https.ClientCertificateMode.DelayCertificate> is new option available in .NET 6 or later. When set, an application can first check the <xref:Microsoft.AspNetCore.Http.ConnectionInfo.ClientCertificate> property to see if the certificate is already available. If it is not available, ensure the request body has been consumed before calling <xref:Microsoft.AspNetCore.Http.ConnectionInfo.GetClientCertificateAsync*> to negotiate one. Note `GetClientCertificateAsync` can return a null certificiate if the client declines to provide one.

Reviewer: ```suggestion
<!-- TODO when API is published <xref:Microsoft.AspNetCore.Server.Kestrel.Https.ClientCertificateMode.DelayCertificate> 
-->
`Kestrel.Https.ClientCertificateMode.DelayCertificate` is new option available in .NET 6 or later. When set, an app can check the <xref:Microsoft.AspNetCore.Http.ConnectionInfo.ClientCertificate> property to see if the certificate is available. If it isn't available, ensure the request body has been consumed before calling <xref:Microsoft.AspNetCore.Http.ConnectionInfo.GetClientCertificateAsync*> to negotiate one. Note `GetClientCertificateAsync` can return a null certificate if the client declines to provide one.
```

*NOTE* The application should buffer or consume any request body data before attempting the renegotiation, otherwise `GetClientCertificateAsync` may throw `InvalidOperationExeption: Client stream needs to be drained before renegotiation.`.

> If you are programatically configuring the TLS settings per host there is a new [UseHttps](xref:fundamentals/servers/kestrel/endpoints#listenoptionsusehttps) overload available in .NET 6 and later that takes <xref:Microsoft.AspNetCore.Server.Kestrel.Https.TlsHandshakeCallbackOptions> and controls client certificate renegotiation via <xref:Microsoft.AspNetCore.Server.Kestrel.Https.TlsHandshakeCallbackContext.AllowDelayedClientCertificateNegotation>.

Reviewer: ```suggestion
If you're programmatically configuring the TLS settings per host there is a new [UseHttps](xref:fundamentals/servers/kestrel/endpoints#listenoptionsusehttps) overload available in .NET 6 and later that takes `Server.Kestrel.Https.TlsHandshakeCallbackOptions` and controls client certificate renegotiation via `Kestrel.Https.TlsHandshakeCallbackContext.AllowDelayedClientCertificateNegotation`.

<!-- TODO .NET 6 API <xref:Microsoft.AspNetCore.Server.Kestrel.Https.TlsHandshakeCallbackOptions>
<xref:Microsoft.AspNetCore.Server.Kestrel.Https.TlsHandshakeCallbackContext.AllowDelayedClientCertificateNegotation>
 -->
```
